src/code42cli/cmds/devices.py

```diff
@@ -0,0 +1,368 @@
+        }
+        if macos_guid:
+            try:
+                full_disk_access_status = sdk.devices.get_agent_full_disk_access_state(guid).data["value"] # returns 404 error if device isn't a Mac or doesn't have full disk access
+            except:
+                full_disk_access_status = False
+        else:
+            full_disk_access_status=""
+        current_result_dict["full_disk_access"] = full_disk_access_status
+        return current_result_dict
+
+    result_list = run_bulk_process(handle_row, rows, progress_label="Getting device settings")
+    return device_dataframe.merge(DataFrame.from_records(result_list), on="guid")
+
+def _drop_devices_which_have_not_connected_in_some_number_of_days(
+    devices_dataframe, days_since_last_connected
+):
+    return devices_dataframe.loc[
+        to_datetime(datetime.now(), utc=True)
+        - to_datetime(devices_dataframe["lastConnected"], utc=True)
+        > to_timedelta(days_since_last_connected, unit="days"),
+        :,
+    ]
+
+
+def _drop_n_devices_per_user(
+    device_dataframe,
+    number_to_drop,
+    sort_field="lastConnected",
+    sort_ascending=False,
+    group_field="userUid",
+):
+    return (
+        device_dataframe.sort_values(by=sort_field, ascending=sort_ascending)
+        .drop(device_dataframe.groupby(group_field).head(number_to_drop).index)
+        .reset_index(drop=True)
+    )
+
+
+def _add_usernames_to_device_dataframe(sdk, device_dataframe):
+    users_generator = sdk.users.get_all()
+    users_list = []
+    for page in users_generator:
+        users_list.extend(page["users"])
+    users_dataframe = DataFrame.from_records(
+        users_list, columns=["username", "userUid"]
+    )
+    return device_dataframe.merge(users_dataframe, how="left", on="userUid")
+
+
+@bulk.command(
+    name="deactivate",
+    help="Deactivate all devices on the given list. Takes as input a CSV with a deviceId column",
+)
+@read_csv_arg(headers=["deviceId"])
+@change_device_name_option
+@purge_date_option
+@format_option
+@sdk_options()
+def bulk_deactivate(state, csv_rows, change_device_name, purge_date, format):
+    sdk = state.sdk
+    csv_rows[0]["deactivated"] = False
+    formatter = OutputFormatter(format, {key: key for key in csv_rows[0].keys()})
+    for row in csv_rows:
+        row["change_device_name"] = change_device_name
+        row["purge_date"] = purge_date
+
+    def handle_row(**row):
+        try:
+            _deactivate_device(
+                sdk, row["deviceId"], row["change_device_name"], row["purge_date"]
+            )
+            row["deactivated"] = "True"
+        except Exception as e:
+            row["deactivated"] = "False: {}".format(e)
+        return row
+
+    result_rows = run_bulk_process(
+        handle_row, csv_rows, progress_label="Deactivating devices:"
+    )
+    click.echo("Run complete.")
```

Comment:
Is this needed, intentionally printing the statement?

Reply:
Nope, removed.